src/Functions/isIPAddressContainedIn.cpp

```diff
@@ -0,0 +1,739 @@
+            uint8_t addr[16];
+            friend class Address;
+        };
+
+        constexpr const std::variant<V4, V6> & variant() const
+        {
+            return addr;
+        }
+
+    private:
+        std::variant<V4, V6> addr;
+    };
+
+    class CIDR
+    {
+    public:
+        CIDR() = delete;
+
+        explicit CIDR(const StringRef & in)
+        {
+            const auto in_view = std::string_view(in);
+            const auto pos_slash = in_view.find('/');
+
+            if (pos_slash == std::string_view::npos)
+                throw DB::Exception("The text does not contain '/': " + std::string(in_view),
+                                    DB::ErrorCodes::CANNOT_PARSE_TEXT);
+
+            prefix = Address(StringRef(in_view.substr(0, pos_slash)));
+
+            // DB::parse<Uint8>() in <IO/ReadHelpers.h> ignores
+            // non-digit characters. std::stoi() skips whitespaces. We
+            // need to parse the prefix bits in a strict way.
+
+            if (pos_slash + 1 == in_view.size())
+                throw DB::Exception("The CIDR has no prefix bits: " + std::string(in_view),
+                                    DB::ErrorCodes::CANNOT_PARSE_TEXT);
+
+            bits = 0;
+            for (size_t i = pos_slash + 1; i < in_view.size(); i++)
+            {
+                const auto c = in_view[i];
+                if (c >= '0' && c <= '9')
+                {
+                    bits *= 10;
+                    bits += c - '0';
+                }
+                else
+                {
+                    throw DB::Exception("The CIDR has a malformed prefix bits: " + std::string(in_view),
+                                        DB::ErrorCodes::CANNOT_PARSE_TEXT);
+                }
+            }
+
+            if ((std::holds_alternative<Address::V4>(prefix->variant()) && bits > 32) || bits > 128)
+                throw DB::Exception("The CIDR has an invalid prefix bits: " + std::string(in_view),
+                                    DB::ErrorCodes::CANNOT_PARSE_TEXT);
+        }
+
+    private:
+        template <typename PrefixT>
+        static PrefixT toMask(uint8_t bits)
+        {
+            if constexpr (std::is_same_v<PrefixT, Address::V4>)
+            {
+                if (bits == 32)
+                    return PrefixT({0xFF, 0xFF, 0xFF, 0xFF});
+                else if (bits >= 24)
+                    return PrefixT({0xFF, 0xFF, 0xFF, static_cast<uint8_t>(0xFF << (8 - bits % 8))});
+                else if (bits >= 16)
+                    return PrefixT({0xFF, 0xFF, static_cast<uint8_t>(0xFF << (8 - bits % 8)), 0x00});
+                else if (bits >= 8)
+                    return PrefixT({0xFF, static_cast<uint8_t>(0xFF << (8 - bits % 8)), 0x00, 0x00});
+                else
+                    return PrefixT({static_cast<uint8_t>(0xFF << (8 - bits % 8)), 0x00, 0x00, 0x00});
+            }
+            else
+            {
+                if (bits == 128)
```

Comment:
We have function to convert integer prefix length to binary representation
https://github.com/ClickHouse/ClickHouse/blob/e27715e55e471dc4d7f825eb4097b468a9a4a451/src/Common/IPv6ToBinary.cpp#L67

It statically generate all of 128 possible masks. But maybe this implementation with `if (bits >= ...) ... else if ...` event better, not sure. Anyway, we can leave one implementation.

Reply:
Okay, I will reuse that function. Thanks for the feedback.